From 8c29f0f52f3265e8d503f02f88aabeb06c736c0d Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Sun, 21 May 2017 10:54:16 -0400
Subject: [PATCH] Fix a crash in jp2 codec

Lack of validation of jp2 could lead to a crash

The patch is a merge of 3 upstream commit

bug-debian: https://bugs.debian.org/869830
bug; https://github.com/ImageMagick/ImageMagick/issues/501
origin: https://github.com/ImageMagick/ImageMagick/commit/acee073df34aa4d491bf5cb74d3a15fc80f0a3aa,
 https://github.com/ImageMagick/ImageMagick/commit/ac23b02ecb741e5de60f5235ea443790c88a0b80,
 https://github.com/ImageMagick/ImageMagick/commit/b0c5222ce31e8f941fa02ff9c7a040fb2db30dbc
(cherry picked from commit acee073df34aa4d491bf5cb74d3a15fc80f0a3aa,
ac23b02ecb741e5de60f5235ea443790c88a0b80, b0c5222ce31e8f941fa02ff9c7a040fb2db30dbc)
---
 coders/jp2.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/coders/jp2.c b/coders/jp2.c
index c997c283d..154a0ac55 100644
--- a/coders/jp2.c
+++ b/coders/jp2.c
@@ -388,7 +388,12 @@ static Image *ReadJP2Image(const ImageInfo *image_info,ExceptionInfo *exception)
   opj_stream_destroy(jp2_stream);
   for (i=0; i < (ssize_t) jp2_image->numcomps; i++)
   {
-    if ((jp2_image->comps[i].dx == 0) || (jp2_image->comps[i].dy == 0))
+    if ((jp2_image->comps[0].dx == 0) || (jp2_image->comps[0].dy == 0) ||
+        (jp2_image->comps[0].dx != jp2_image->comps[i].dx) ||
+        (jp2_image->comps[0].dy != jp2_image->comps[i].dy) ||
+        (jp2_image->comps[0].prec != jp2_image->comps[i].prec) ||
+        (jp2_image->comps[0].sgnd != jp2_image->comps[i].sgnd) ||
+        ((image->ping == MagickFalse) && (jp2_image->comps[i].data == NULL)))
       {
         opj_destroy_codec(jp2_codec);
         opj_image_destroy(jp2_image);
